File: transpiler/javatests/com/google/j2cl/readable/java/protobuf/Protobuf.java
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package protobuf;

import com.google.j2objc.annotations.ObjectiveCName;
import com.google.protobuf.ExtensionLite;
import com.google.protobuf.GeneratedMessage;
import com.google.protobuf.Parser;
import org.jspecify.annotations.NullMarked;

@NullMarked
public class Protobuf {
  public void test() {
    MyMessage.Builder builder = MyMessage.newBuilder();

    builder.getFoo();
    builder.setFoo(1);

    builder.getFooBar();
    builder.setFooBar(2);

    builder.getX();
    builder.setX(3);

    builder.getXValue();
    builder.setXValue(4);

    builder.getXYZ();
    builder.setXYZ(5);

    builder.getXYZValues();
    builder.setXYZValues(6);

    MyMessage message = builder.build();
    message.getFoo();
    message.getFooBar();
    message.getX();
    message.getXValue();
    message.getXYZ();
    message.getXYZValues();

    MyMessage chainedMessage =
        MyMessage.newBuilder()
            .setFoo(1)
            .setFooBar(2)
            .setX(3)
            .setXValue(4)
            .setXYZ(5)
            .setXYZValues(6)
            .build();

    MyMessage defaultMessage = MyMessage.getDefaultInstance();
    defaultMessage.getFoo();

    Parser<?> parser = message.getParserForType();
    Parser<? extends GeneratedMessage> generatedMessageParser =
        ((GeneratedMessage) message).getParserForType();

    MyEnum enumOne = MyEnum.ONE;
    int enumOneNumber = MyEnum.ONE.getNumber();
  }

  public <T> void testExtension(ExtensionLite<T> extension) {
    int i = extension.getNumber();
  }

  public static class Test {
    public Test(MyMessage myMessage) {}

    public void accept(MyMessage myMessage) {}

    public static void acceptStatic(MyMessage myMessage) {}

    @ObjectiveCName("acceptStaticWithMyMessage:")
    public static void acceptStaticWithObjectiveCName(MyMessage myMessage) {}

    public MyMessage get() {
      throw new RuntimeException();
    }

    public static MyMessage getStatic(MyMessage myMessage) {
      throw new RuntimeException();
    }

    @ObjectiveCName("getStaticWithMyMessage:")
    public static MyMessage getStaticWithObjectiveCName(MyMessage myMessage) {
      throw new RuntimeException();
    }
  }
}

File: transpiler/java/com/google/j2cl/transpiler/backend/kotlin/Protobuf.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.j2cl.transpiler.backend.kotlin

import com.google.j2cl.transpiler.ast.MethodDescriptor
import com.google.j2cl.transpiler.ast.TypeDeclaration
import com.google.j2cl.transpiler.backend.kotlin.common.camelCaseStartsWith
import com.google.j2cl.transpiler.backend.kotlin.common.runIf

internal val TypeDeclaration.isProtobuf: Boolean
  get() = allSuperTypesIncludingSelf.any { it.packageName == "com.google.protobuf" }

internal val MethodDescriptor.isProtobuf: Boolean
  get() = !isStatic && enclosingTypeDescriptor.typeDeclaration.isProtobuf

private const val GET: String = "get"

internal val MethodDescriptor.isProtobufGetter: Boolean
  get() = isProtobuf && parameterDescriptors.isEmpty() && name?.startsWith(GET) ?: false

internal fun String.toProtobufPropertyName() =
  runIf(camelCaseStartsWith(GET)) { substring(GET.length).replaceFirstChar { it.toLowerCase() } }
